zebra-rpc/src/methods.rs

```diff
@@ -613,7 +615,14 @@ where
                 match response {
                     zebra_state::ReadResponse::TransactionIdsForBlock(Some(tx_ids)) => {
                         let tx_ids = tx_ids.iter().map(|tx_id| tx_id.encode_hex()).collect();
-                        Ok(GetBlock::Object { tx: tx_ids })
+                        Ok(GetBlock::Object {
+                            hash: block.hash(),
+                            height: block
+                                .coinbase_height()
+                                .expect("all blocks should have a coinbase height"),
+                            size: block.zcash_serialized_size(),
```

Comment:
Reading entire block when `verbose = 1` will cause performance issues for `lightwalletd`.

`lightwalletd` calls the `getblock` RPC twice for every block during its initial sync, first with `verbose = 1` to get the transaction IDs, and again with `verbose = 0` to get the block data:
https://github.com/zcash/lightwalletd/blob/5d174f7feb702dc19aec5b09d8be8b3d5b17ce45/common/common.go#L246-L254

This caused a serious performance issue for Zebra, here's how we fixed it a few months ago:
- https://github.com/ZcashFoundation/zebra/issues/5298#issuecomment-1262775222
- https://github.com/ZcashFoundation/zebra/pull/5307

(I should have added some comments about this when I fixed it, sorry about that!)

When we were fetching the full block when `verbose = 1`, a full `lightwalletd` sync took 7 hours. (Compared with 3.4 hours after the fix in September.)

`zcashd` takes around 5 hours for a full `lightwalletd` sync. This change would make us much slower than `zcashd`.

This seems like a blocker, because one of our release goals is being faster than `zcashd`.

### Suggested Fix

We want Zebra to avoid reading the whole block for `lightwalletd`, which needs to efficiently get the `tx` field. But `zebra-checkpoints` needs the `hash`, `height`, and `size` fields.

Here are some ideas:
1. We could add a private `verbose` mode for `zebra-checkpoints`. Then `lightwalletd` can continue using `verbose=1`, and we can make any changes we want to the new mode.
2. Or we can add a separate private RPC with just the fields that `zebra-checkpoints` needs.

When `zebra-checkpoints` starts up, it can check if it's connected to Zebra, and use the private mode/RPC.

This is all temporary, until either:
- `lightwalletd` gets fixed so it calculates v5 transaction IDs correctly
- Zebra stores the block size in the database (ticket #5911)

What do you think?

### Other Alternatives

#### Wait for Fixes

That seems like a bit of work, and it's a bit complicated. So we could wait for the proper fixes in Zebra's state, or in `lightwalletd`. We could update the ticket with the progress we've made, and a list of blockers?

#### Get Fields Efficiently

I thought of some ways to get the RPC fields efficiently. But I couldn't get all of them without reading the block.

- hash: Call `ReadRequest::BestChainBlockHash(height)` or if the hash was supplied, use it
    - we'll need to change this request so it is compiled unconditionally, by removing the `getblocktemplate-rpcs` feature gate
- height: Add a new `ReadRequest::BestChainBlockHeight(hash)`, or if the height was supplied, use it
    - you can use `read::height_by_hash()` to do this efficiently
- size: RocksDB doesn't store the size of each value, it reads the disk instead. So we'd need to add the serialized block size to the database to do this efficiently (probably in `hash_by_height`).
    - this seems out of scope for this PR, so I opened another ticket #5911

#### Redesign zebra-checkpoints

We could use the number of transactions for checkpoint block spacing, rather than the number of bytes in each block. We already get this efficiently from `verbose=1`.

 I'm not sure what impact this would have on Zebra's sync performance, because transactions can be very different sizes.

Reply:
> lightwalletd gets fixed so it calculates v5 transaction IDs correctly

Is there an issue for that? I don't see it at https://github.com/zcash/lightwalletd/issues .

Reply:
There was an issue:
https://github.com/zcash/lightwalletd/issues/392

But it was closed with this workaround:
https://github.com/zcash/lightwalletd/pull/393

```go
	// `block.ParseFromSlice` correctly parses blocks containing v5
	// transactions, but incorrectly computes the IDs of the v5 transactions.
	// We temporarily paper over this bug by fetching the correct txids via a
	// verbose getblock RPC call, which returns the txids.
	//
	// Unfortunately, this RPC doesn't return the raw hex for the block,
	// so a second getblock RPC (non-verbose) is needed (below).
```
https://github.com/zcash/lightwalletd/blob/07fd6a966a136a751b37b8c54a6679499def0caa/common/common.go#L246

I'm not sure how that team handles technical debt.

It is likely only a performance issue for `lightwalletd`'s initial sync, because the slowdown is much smaller than the target block interval.

Reply:
We could also redesign zebra-checkpoints to call `getblock` with `verbose=0` and deserialize the response or only return height/hash/size from `verbose=1` when compiled without the lightwalletd feature.

Reply:
> We could also redesign zebra-checkpoints to call `getblock` with `verbose=0` and deserialize the response,

That's a great idea! Let's do it.

It will be a bit slow to download all the block data, but we only update the checkpoints every few weeks, so that's ok.

As optional performance improvements, we could:
* keep the existing `getblock verbose=1` code, and use it with `zcashd`. Then `zcashd` will stay the same speed.
* only deserialize the block header in Zebra mode, then just count the rest of the bytes to get the block size.

> or only return height/hash/size from `verbose=1` when compiled without the lightwalletd feature.

Unfortunately, there is no lightwalletd feature in Zebra. (And Rust features are meant to be additive.) So I'm not sure if this alternative would work. But the other one seems promising.

Reply:
I don't see how deserializing the header only will give you the hash and height. Deserializing the entire block should work, with the known performance problem. I think i will try that.